lighthouse-core/lib/asset-saver.js

```diff
@@ -103,6 +104,68 @@ function prepareAssets(artifacts, audits) {
 }
 
 /**
+ * Generates a JSON representation of traceData line-by-line to avoid OOM due to
+ * very large traces.
+ * @param {{traceEvents: !Array}} traceData
+ * @return {!Iterator<string>}
+ */
+function* traceJsonGenerator(traceData) {
+  const keys = Object.keys(traceData);
+
+  yield '{\n';
+
+  // Stringify and emit trace events separately to avoid a giant string in memory.
+  yield '"traceEvents": [\n';
+  if (traceData.traceEvents.length > 0) {
+    const eventsIterator = traceData.traceEvents[Symbol.iterator]();
+    // Emit first item manually to avoid a trailing comma.
+    const firstEvent = eventsIterator.next().value;
+    yield `  ${JSON.stringify(firstEvent)}`;
+    for (const event of eventsIterator) {
+      yield `,\n  ${JSON.stringify(event)}`;
+    }
+  }
+  yield '\n]';
+
+  // Emit the rest of the object (usually just `metadata`)
+  if (keys.length > 1) {
+    for (const key of keys) {
+      if (key === 'traceEvents') continue;
+
+      yield `,\n"${key}": ${JSON.stringify(traceData[key], null, 2)}`;
+    }
+  }
+
+  yield '}\n';
+}
+
+/**
+ * Save a trace as JSON by streaming to disk at traceFilename.
+ * @param {{traceEvents: !Array}} traceData
+ * @param {string} traceFilename
+ * @return {!Promise<undefined>}
+ */
+function saveTrace(traceData, traceFilename) {
+  return new Promise((resolve, reject) => {
+    const traceIter = traceJsonGenerator(traceData);
+    // A stream that pulls in the next traceJsonGenerator chunk as writeStream
+    // reads from it. Closes stream with null when iteration is complete.
+    const traceStream = new stream.Readable({
+      read() {
+        const next = traceIter.next();
+        this.push(next.done ? null : next.value);
+      }
+    });
+
+    const writeStream = fs.createWriteStream(traceFilename);
```

Comment:
interesting idea, through `Buffer.concat`? We could check it out, though assembling an entire giant trace in memory makes me nervous since of course the original copy of the giant trace is also taking up memory :)

We could also do a `fs.writeSync` loop, but it didn't seem to really matter since `saveAssets` is already async due to screenshots.